### PR TITLE
[epoch] Emergency Framework Upgrades

### PR DESCRIPTION
**Pull request by amnn**
First commit introduces a special escape hatch to inject new versions of Sui Framework at pre-determined epoch boundaries, by overwriting the package and then pushing the node over with a panic.

## Test Plan

The second commit demonstrates how to create an overridden sui-framework by:

- Introducing a new event type `sui::sui::Canary` and a new entry function, `sui::sui::coal_mine` which emits it.
- Copying `crates/sui-framework` to `crates/sui-framework-override` along with the new function, however the event emits  a different value.
- Setting an override for the framework on the boundary between epochs 4 and 5.

Can be tested locally by running a validator:

```
sui$ cargo build
sui$ rm -rf ~/.sui && ~/sui/target/debug/sui genesis -f
# Set checkpoints-per-epoch, to a smaller value in ~/.sui/sui_config/network.yaml (I chose 100)
sui$ ~/sui/target/debug/sui start --no-full-node
```

And a fullnode in another session:
```
sui$ ~/sui/target/debug/sui-node --config-path ~/.sui/sui_config/fullnode.yaml
```

And faucet in yet another session:
```
sui$ ~/sui/target/debug/sui-faucet --write-ahead-log /tmp/faucet.wal
```

And sending requests in yet another sesson:
```
sui$ ~/sui/target/debug/sui client call \
  --package "0x2" --module "sui" --function "coal_mine" --gas-budget 10000
```

When this request succeeds before the emergency upgrade, the event contains value 42, and when it succeeds after, it contains value 43.  When the upgrade epoch is hit, both the validators and fullnode will need to be restarted.

### CODE REVIEW

File: crates/sui-node/src/lib.rs [hunk 1]
```diff
@@ -689,6 +692,47 @@ impl SuiNode {
...
+
+        if curr_package == new_package {
+            info!(next_epoch, "Emergency Upgrade - picking up upgraded package");
+            return;
+        }
+
+        let package_ref = curr_package.compute_object_reference();
+        match store.insert_object_direct(package_ref, &new_package).await {
+            Ok(()) => {
+                panic!("Emergency Upgrade - at epoch {next_epoch}, please restart.")
```
Comment (sblackshear) [inline crates/sui-node/src/lib.rs:726]: Capturing the nice Slack explanation of why this works in a comment would be useful here.

**Review by lxfind — COMMENTED** [review]
Curious what's the trade-off between embedding the new framework as source code or as binary

**Comment by amnn** [comment]
The main pro to using bytecode is that we don't include the move compiler in the validator codebase, because it's not intended to be part of the trusted compute base, the secondary pro is that storing the override as bytecode means we get to reuse the build mechanism we have for the original framework.

Main pro I can see for source is that it's maybe easier to audit?  But the trusted compute base part is a pretty compelling vote for bytecode, for me.

File: crates/sui-config/src/framework.rs [hunk 2]
```diff
@@ -0,0 +1,13 @@
+// Copyright (c) Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+use move_binary_format::CompiledModule;
+
+/// TEMPORARY: Detect if we should update to a new framework before epoch `epoch` starts.  Returns
+/// the modules to replace the Sui Framework with, if an upgrade is detected, or None otherwise.
+pub fn override_sui_framework(epoch: u64) -> Option<Vec<CompiledModule>> {
+    match epoch {
+        5 => Some(sui_framework_override::get_sui_framework()),
```
Comment (mystenmark) [inline crates/sui-config/src/framework.rs:10]: I assume this was here for testing - remember to remove before merging.

**Review by mystenmark — COMMENTED** [review]
Looks good overall, but let's not merge to main - this should only go to the testnet branch, right?

**Comment by amnn** [comment]
@mystenmark, right this should go to testnet (there were just some existing issues with the tests in `testnet` at the time of development, that's why it's based off `main`).  I would say that we should also not land the change (even without the testing commit) until we need to make a Sui Framework Upgrade -- that's why I've kept that test commit in there (to offer a basis for a hypothetical actual framework change).

**Comment by amnn** [comment]
Retiring this PR, now that we've landed framework upgrades for real.

**Pull request closed without merging**